internal/metricsv2/operations_stats.go

```diff
@@ -0,0 +1,231 @@
+	defer func() {
+		if recovery := recover(); recovery != nil {
+			s.logger.Errorf("panic recovered while creating and registering operations metrics: %v", recovery)
+		}
+	}()
+
+	for _, plan := range plans {
+		for _, opType := range opTypes {
+			for _, opState := range opStates {
+				key, err := s.makeKey(opType, opState, plan)
+				setup.FatalOnError(err)
+				name, err := s.buildName(opType, opState)
+				setup.FatalOnError(err)
+				labels := prometheus.Labels{"plan_id": string(plan)}
+				switch opState {
+				case domain.InProgress:
+					s.gauges[key] = prometheus.NewGauge(
+						prometheus.GaugeOpts{
+							Name:        name,
+							ConstLabels: labels,
+						},
+					)
+					prometheus.MustRegister(s.gauges[key])
+				case domain.Failed, domain.Succeeded:
+					s.counters[key] = prometheus.NewCounter(
+						prometheus.CounterOpts{
+							Name:        name,
+							ConstLabels: labels,
+						},
+					)
+					prometheus.MustRegister(s.counters[key])
+				}
+			}
+		}
+	}
+
+	go s.statsFromDB(ctx)
+}
+
+func (s *operationStats) Handler(_ context.Context, event interface{}) error {
+	defer func() {
+		if recovery := recover(); recovery != nil {
+			s.logger.Error("panic recovered while handling operation counting event: %v", recovery)
+		}
+	}()
+
+	payload, ok := event.(process.OperationCounting)
+	if !ok {
+		return fmt.Errorf("expected process.OperationStepProcessed but got %+v", event)
+	}
+
+	opState := payload.OpState
+
+	if opState != domain.Failed && opState != domain.Succeeded {
+		return fmt.Errorf("operation state is %s, but operation counter support events only from failed or succeded operations", payload.OpState)
+	}
+
+	key, err := s.makeKey(payload.OpType, opState, payload.PlanID)
+	if err != nil {
+		s.logger.Error(err)
+		return err
+	}
+
+	metric, found := s.counters[key]
+	if !found || metric == nil {
+		return fmt.Errorf("metric not found for key %s", key)
+	}
+	s.counters[key].Inc()
+
+	return nil
+}
+
+func (s *operationStats) statsFromDB(ctx context.Context) {
+	defer func() {
+		if recovery := recover(); recovery != nil {
+			s.logger.Errorf("panic recovered while handling in progress operation counter: %v", recovery)
+		}
+	}()
+
+	ticker := time.NewTicker(s.poolingInterval)
+	for {
+		select {
+		case <-ticker.C:
+			stats, err := s.operations.GetOperationStatsByPlanV2()
+			if err != nil {
+				s.logger.Errorf("cannot fetch in progress operations from db : %s", err.Error())
+				continue
+			}
+			setStats := make(map[metricKey]struct{})
+			for _, stat := range stats {
+				key, err := s.makeKey(stat.Type, stat.State, broker.PlanID(stat.PlanID))
+				if err != nil {
+					s.logger.Error(err)
+					continue
+				}
+
+				metric, found := s.gauges[key]
+				if !found || metric == nil {
+					s.logger.Errorf("metric not found for key %s", key)
+					continue
+				}
+				metric.Set(float64(stat.Count))
+				setStats[key] = struct{}{}
+			}
+
+			for key, metric := range s.gauges {
```

Comment:
You can start with `s.gauges`, initiate all to 0, then process `stats` and update particular values.
Then `setStats` map would not be needed.